File: This Is Coding Test with Python/Python/ThisIsCodingTest/Greedy/[3-3] 숫자 카드 게임.py
class Param:
    def __init__(self, values: list):
        self.n, self.m, = values[:2]
        self.v = [values[2+n*self.m:2+n*self.m+self.m] for n in range(0, self.n)]


def MySolution(param: Param):
    v = param.v
    mins = list()

    for i in v:
        i.sort()
        mins.append(i[0])

    return max(mins)


def BookSolution(param):
    v = param.v
    result = 0

    for i in v:
        result = max(min(i), result)

    return result


if __name__ == "__main__":
    import CodingTester

    class TestSet:
        def __init__(self, num, p: list, r):
            self.num = num
            self.param: Param
            if p.__class__ is list:
                self.param = Param(p)
            elif p.__class__ is Param:
                self.param = p

            self.result = r

        def __format__(self, format_spec):
            num_str = "[" + str(self.num) + "]"
            a = f" {num_str:4} | n: {self.param.n}, m: {self.param.m}\n"
            b = str()
            for n in range(self.param.n):
                b += f"      | "
                for m in range(self.param.m):
                    b += f"{self.param.v[n][m]} "
                b += "\n"
            c = f"      | result: {self.result}"
            return a + b + c

    file = CodingTester.ReadTestFile("../../../TestSets/3-3.txt")
    test_sets = list()

    i = 0
    while i < len(file):
        test_sets.append(TestSet(0, file[i:i + file[i] * file[i+1] + 2], file[i + file[i] * file[i+1] + 2]))
        i += file[i] * file[i+1] + 3

    for i, test_set in enumerate(test_sets):
        CodingTester.OutputTestSolutionExt(MySolution, i + 1, test_set.param, test_set.result, TestSet)
        print()

    for i, test_set in enumerate(test_sets):
        CodingTester.OutputTestSolutionExt(BookSolution, i + 1, test_set.param, test_set.result, TestSet)
        print()

File: This Is Coding Test with Python/C++/ThisIsCodingTest-Basic/Greedy/[3-3] 숫자 카드 게임.cpp
#include <iostream>
#include <vector>
#include "CodingTester.h"

using namespace std;

struct Param {
	int n{};
	int k{};
	vector<vector<int>> v{};

	friend istream& operator>>( istream& is, Param& self )
	{
		is >> self.n >> self.k;

		int elm{};
		for ( int n{}; n < self.n; ++n ) {
			self.v.push_back( {} );
			for ( int m{}; m < self.k; ++m ) {
				is >> elm;
				self.v[n].emplace_back(elm);
			}
		}

		return is;
	}
};

struct TestSet {
	int num{};
	Param param{};
	int result{};

	TestSet() = default;
	TestSet( Param p, int r ) {
		TestSet();
		param = p;
		result = r;
	}
	friend istream& operator>>( istream& is, TestSet& t )
	{
		is >> t.param;
		is >> t.result;

		return is;
	}
};

template <>
struct std::formatter<TestSet> {
	constexpr auto parse( format_parse_context& ctx ) { return ctx.begin(); }

	template <typename FormatContext>
	auto format( const TestSet& ts, FormatContext& ctx ) {

		string strnum = "[" + to_string( ts.num ) + "]";
		auto out = format_to( ctx.out(), " {:4} | ", strnum );

		out = format_to( out, "N:{}, M:{}\n", ts.param.n, ts.param.k );

		for ( int n{}; n < ts.param.n; ++n ) {
		out = format_to( out, "{:5} | ", "" );

			for ( int m{}; m < ts.param.k; ++m ) {
				out = format_to( out, "{} ", ts.param.v[n][m] );
			}
			out = format_to( out, "\n" );
		}

		out = format_to( out, "{:5} | ", "" );
		out = format_to( out, "Result: {}", ts.result );

		return out;
	}
};

int MySolution( Param param );
int BookSolution( Param param );

int main()
{
	auto test_sets{ ReadTestFile<TestSet>( "../../../TestSets/3-3.txt" ) };

	for ( int i{}; const auto & test_set : test_sets ) {
		OutputTestSolution<Param, int, TestSet>( MySolution, ++i, test_set.param, test_set.result );
		cout << endl;
	}

	for ( int i{}; const auto & test_set : test_sets ) {
		OutputTestSolution<Param, int, TestSet>( BookSolution, ++i, test_set.param, test_set.result );
		cout << endl;
	}
}

int MySolution( Param param )
{
	auto v = param.v;
	vector<int> mins{};

	for ( auto elm : v )
		mins.emplace_back( *min_element(elm.begin(), elm.end()) );
	
	return *max_element( mins.begin(), mins.end() );
}

int BookSolution( Param param )
{
	auto v = param.v;
	int result{};

	for ( auto elm : v )
		result = max( *min_element( elm.begin(), elm.end() ), result );

	return result;
}
